### PR TITLE
Add support for google::protobuf::Repeated(Ptr)Field in Collect()

### PR DESCRIPTION
**Pull request by FolMing**
Addresses #477.

### CODE REVIEW

File: eventuals/collect.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 #include "eventuals/loop.h"
 #include "eventuals/type-traits.h"
+#include "google/protobuf/repeated_field.h"
```
Comment (CodingCanuck) [inline eventuals/collect.h:7]: I'm confused: how does this `#include` succeed? I don't see any dependencies on something like `com_google_protobuf`.

Ah, I'm guessing we're pulling this in from an indirect dependency on `stout`: https://github.com/3rdparty/stout-flags/blob/fdff8dd353991afc193f77506f04da505237a796/bazel/repos.bzl

Can you please add a direct dependency on a relevant proto target to this file's build rule so the source of the dependency is more clear?

File: eventuals/collect.h [hunk 2]
```diff
@@ -16,7 +17,23 @@ template <typename Container>
   return Loop<Container>()
       .context(Container())
       .body([](auto& data, auto& stream, auto&& value) {
-        if constexpr (HasEmplaceBack<Container>::value) {
+        if constexpr (
+            std::is_same_v<
+                Container,
+                google::protobuf::RepeatedField<
+                    typename Container::value_type>>) {
+          data.Add(value);
```
Comment (CodingCanuck) [inline eventuals/collect.h:25]: std::move(value)

File: test/collect.cc [hunk 3]
```diff
@@ -39,5 +39,35 @@ TEST(Collect, CommonSetPass) {
...
+  std::vector<std::string> v = {"Hello", "World"};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | Collect<google::protobuf::RepeatedPtrField<std::string>>();
+  };
+
+  google::protobuf::RepeatedPtrField<std::string> result = *s();
+
+  EXPECT_EQ(result.size(), 2);
```
Comment (CodingCanuck) [inline test/collect.cc:52]: Prefer `ASSERT` over `EXPECT` when checking a container's size before dereferencing elements, or doing other things that could crash if previous `EXPECT` statements fail.

> Use ASSERT when it doesn't make sense to continue

https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

File: test/collect.cc [hunk 4]
```diff
@@ -39,5 +39,35 @@ TEST(Collect, CommonSetPass) {
...
+  EXPECT_STREQ("Hello", result.begin()->c_str());
```
Comment (CodingCanuck) [inline test/collect.cc:53]: Whenever you're passing `foo->c_str()` to `EXPECT_STREQ`, use `EXPECT_EQ` instead:

`EXPECT_EQ("Hello", *result.begin());`

Here and below.

See discussion: https://github.com/3rdparty/stout/pull/53#discussion_r903157994

File: eventuals/collect.h (hunk 1 above)
Comment (CodingCanuck) [inline eventuals/collect.h:7]: This feels like a pretty heavy dependency for users that aren't using protos.

Is there any way we can avoid it, e.g. by splitting out the proto parts into a separate file?

I have no idea how to do this, so if I was writing this PR I'd have this same `#include`, even though it feels messy :/

Maybe add a TODO to figure out a cleaner way to separate out proto overrides from proto-agnostic stuff?

**Review by CodingCanuck — APPROVED** [review]

File: eventuals/collect.h (hunk 2 above)
Comment (FolMing) [reply eventuals/collect.h:25]: Add's argument is `const &`, so no need to move

Comment (CodingCanuck) [reply eventuals/collect.h:25]: TIL!

File: eventuals/collect.h (hunk 1 above)
Comment (FolMing) [reply eventuals/collect.h:7]: Done, please review the files again and resolve the conversation if everything is ok!

Comment (FolMing) [reply eventuals/collect.h:7]: I'm not sure if we want a TODO here. It might be better to make an issue for that.

Comment (CodingCanuck) [reply eventuals/collect.h:7]: Looks excellent, thanks!

Comment (CodingCanuck) [reply eventuals/collect.h:7]: why not both

I think of TODOs kind of like hyperlinks to issues, plus they give the reader more context (like normal contexts). Personally, I add them liberally :)

Up to you, though.

**Review by CodingCanuck — APPROVED** [review]

**Review by rjhuijsman — APPROVED** [review]
Awesome! Thanks! 🎉

File: eventuals/collect.h (hunk 1 above)
Comment (FolMing) [reply eventuals/collect.h:9]: Added a simple note. If we encounter any problems with compilation times while including `eventuals/collect.h` we'll at least see "HERE'S YOUR PROBLEM" thingy :D

**Pull request merged**